compression.go

```diff
@@ -0,0 +1,71 @@
+package kafka
+
+const (
+	CompressionNone int8 = iota
+	CompressionGZIP
+	CompressionSnappy
+	CompressionLZ4
+)
+
+var codecs map[int8]CompressionCodec
+
+// RegisterCompressionCodec registers a compression codec so it can be used by a Writer.
+func RegisterCompressionCodec(codec func() CompressionCodec) {
+	if codecs == nil {
+		codecs = make(map[int8]CompressionCodec)
+	}
+
+	c := codec()
+	codecs[c.Code()] = c
+}
+
+// CompressionCodec represents a compression codec to encode and decode
+// the messages.
+// See : https://cwiki.apache.org/confluence/display/KAFKA/Compression
+type CompressionCodec interface {
+	Code() int8
+	Encode(dst, src []byte) (int, error)
+	Decode(dst, src []byte) (int, error)
+}
+
+const compressionCodecMask int8 = 0x03
+const DefaultCompressionLevel int = -1
+
+func init() {
+	RegisterCompressionCodec(func() CompressionCodec {
+		return CompressionCodecNone{}
+	})
+}
+
+type CompressionCodecNone struct{}
```

Comment:
Is this type necessary? Always applying Encode/Decode on messages that have no compression means we're going to double the number of dynamic memory allocations and require a copy of the message values that was not necessary before.

Having a way to represent "no compression" is important but we should just get away with not encoding/decoding in this case so we don't pay an extra performance penalty when no compression is applied.